### PR TITLE
[ART-2816]: New implementation for building rpms

### PR DESCRIPTION
**Pull request by vfreex**
... and Including upstream git commit hash in RPM NVR.

Introduces a new implementation for building rpms.

Per https://issues.redhat.com/browse/ART-2690, we want to get rid of
`tito` in Doozer rpm build to gain more control over the build process.

This PR:

- provides a new command `beta:rpms:build` for building rpms without
  tito, which can eventually replace `rpms:build` once tested;
- includes upstream git commit hash in RPM NVR. e.g. `openshift-4.8.0-202011111040.p0.git.abcdef1.el8`;
- uses asyncio as much as possible to run parallel tasks;
- refactors some code to reuse some existing logic;
- adds more unit tests.

This PR makes a few changes to the existing code, but it shouldn't break
the existing implementation.

## Test builds:
https://brewweb.engineering.redhat.com/brew/taskinfo?taskID=36010125

### CODE REVIEW

**Comment by jupierce** [comment]
@vfreex. This is great! Now that we have the ability, I'd personally like to see rpms:rebase & rpms:build separate (we could then combine with an rpms:rebase-and-build verb if we want, similar to olm-bundle:rebase-and-build). 
The reason being that by separating things out, we can do functional testing on the rebase by itself just as we have functional tests for image rebases now.

**Comment by vfreex** [comment]
>  I'd personally like to see rpms:rebase & rpms:build separate (we could then combine with an rpms:rebase-and-build verb if we want, similar to olm-bundle:rebase-and-build).The reason being that by separating things out, we can do functional testing on the rebase by itself just as we have functional tests for image rebases now.

That's a good point. It should be easy to split it into 2 verbs since internally it is done by 2 functions. Some fields in RPMMetadata are populated during rebase, we just need to do that before build. I'll make an update later.

File: doozerlib/dblib.py [hunk 1]
```diff
@@ -107,8 +107,8 @@ def check_missing_db_env_var(self):
         """
 
         if not (constants.DB_USER in os.environ and constants.DB_PWD in os.environ):
-            self.runtime.logger.error("Environment variables required for db operation missing. Doozer will be running"
-                                      "in no DB use mode.")
+            self.runtime.logger.warning("Environment variables required for db operation missing. Doozer will be running"
+                                        "in no DB use mode.")
```
Comment (sosiouxme) [inline doozerlib/dblib.py:111]: the change is fine, but it's better to make a separate commit for unrelated nits like this.

File: doozerlib/distgit.py [hunk 2]
```diff
@@ -280,6 +286,35 @@ def commit(self, commit_message, log_diff=False):
             assertion.success(rc, "Failure fetching commit SHA for {}".format(self.distgit_dir))
         return sha.strip()
 
+    def push(self):
+        with Dir(self.distgit_dir):
+            self.logger.info("Pushing distgit repository %s", self.name)
+            try:
+                # When initializing new release branches, an large amount of data needs to
```
Comment (sosiouxme) [inline doozerlib/distgit.py:293]: nit
```suggestion
                # When initializing new release branches, a large amount of data needs to
```

File: doozerlib/rpm_builder.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+from doozerlib import brew, exectools
+from doozerlib.distgit import RPMDistGitRepo
+from doozerlib.model import Missing
+from doozerlib.rpmcfg import RPMMetadata
+from doozerlib.runtime import Runtime
+from doozerlib.util import is_in_directory
+
+
+class RPMBuilder:
+    """ It builds RPMs!
```
Comment (sosiouxme) [inline doozerlib/rpm_builder.py:22]: :rofl:

File: doozerlib/runtime.py [hunk 4]
```diff
@@ -1340,7 +1340,7 @@ def parallel_exec(self, f, args, n_threads=None):
             while not ret.ready():
                 ret.wait(60)
         except KeyboardInterrupt:
-            self.logger.warn('SIGINT received, signaling threads to terminate...')
+            self.logger.warning('SIGINT received, signaling threads to terminate...')
```
Comment (sosiouxme) [inline doozerlib/runtime.py:1343]: `%$#@$%@` was this blocking some interrupts? or just making them confusing?

File: doozerlib/rpmcfg.py [hunk 5]
```diff
@@ -72,37 +72,45 @@ def __init__(self, runtime, data_obj, commitish: Optional[str] = None, clone_sou
...
+        self.source_path = None
+        self.source_head = None
+        # path to the specfile
+        self.specfile = None
         if clone_source:
-            self.source_path = self.runtime.resolve_source(self)
-            self.source_head = self.runtime.resolve_source_head(self)
+            self.clone_source()
+
+    def clone_source(self):
```
Comment (sosiouxme) [inline doozerlib/rpmcfg.py:82]: aha! finally one without docs!

**Review by sosiouxme — APPROVED** [review]
thanks @vfreex you are showing us how it is done. i especially appreciate the docs-comments and test cases.

going to merge this. this is fairly invasive, is there any chance that any of these changes interfere with the functioning of non-beta `rpms:build`?
